client/fs_endpoint_test.go

```diff
@@ -0,0 +1,2036 @@
+	frames := make(chan *sframer.StreamFrame, 4)
+	var received []byte
+	go func() {
+		for {
+			frame, ok := <-frames
+			if !ok {
+				return
+			}
+
+			if frame.IsHeartbeat() {
+				continue
+			}
+
+			received = append(received, frame.Data...)
+			if reflect.DeepEqual(received, expected) {
+				close(resultCh)
+				return
+			}
+		}
+	}()
+
+	// Start streaming logs
+	go func() {
+		if err := c.endpoints.FileSystem.logsImpl(
+			context.Background(), false, false, 0,
+			OriginStart, task, logType, ad, frames); err != nil {
+			t.Fatalf("logs() failed: %v", err)
+		}
+	}()
+
+	select {
+	case <-resultCh:
+	case <-time.After(10 * time.Duration(testutil.TestMultiplier()) * streamBatchWindow):
+		t.Fatalf("did not receive data: got %q", string(received))
+	}
+}
+
+func TestFS_logsImpl_Follow(t *testing.T) {
+	t.Parallel()
+
+	c := TestClient(t, nil)
+	defer c.Shutdown()
+
+	// Get a temp alloc dir and create the log dir
+	ad := tempAllocDir(t)
+	defer os.RemoveAll(ad.AllocDir)
+
+	logDir := filepath.Join(ad.SharedDir, allocdir.LogDirName)
+	if err := os.MkdirAll(logDir, 0777); err != nil {
+		t.Fatalf("Failed to make log dir: %v", err)
+	}
+
+	// Create a series of log files in the temp dir
+	task := "foo"
+	logType := "stdout"
+	expected := []byte("012345")
+	initialWrites := 3
+
+	writeToFile := func(index int, data []byte) {
+		logFile := fmt.Sprintf("%s.%s.%d", task, logType, index)
+		logFilePath := filepath.Join(logDir, logFile)
+		err := ioutil.WriteFile(logFilePath, data, 777)
+		if err != nil {
+			t.Fatalf("Failed to create file: %v", err)
+		}
+	}
+	for i := 0; i < initialWrites; i++ {
+		writeToFile(i, expected[i:i+1])
+	}
+
+	// Start the reader
+	firstResultCh := make(chan struct{})
+	fullResultCh := make(chan struct{})
+	frames := make(chan *sframer.StreamFrame, 4)
+	var received []byte
+	go func() {
+		for {
+			frame, ok := <-frames
+			if !ok {
+				return
+			}
+
+			if frame.IsHeartbeat() {
+				continue
+			}
+
+			received = append(received, frame.Data...)
+			if reflect.DeepEqual(received, expected[:initialWrites]) {
+				close(firstResultCh)
+			} else if reflect.DeepEqual(received, expected) {
+				close(fullResultCh)
+				return
+			}
+		}
+	}()
+
+	// Start streaming logs
+	go c.endpoints.FileSystem.logsImpl(
+		context.Background(), true, false, 0,
+		OriginStart, task, logType, ad, frames)
+
+	select {
+	case <-firstResultCh:
+	case <-time.After(10 * time.Duration(testutil.TestMultiplier()) * streamBatchWindow):
+		t.Fatalf("did not receive data: got %q", string(received))
+	}
+
+	// We got the first chunk of data, write out the rest to the next file
+	// at an index much ahead to check that it is following and detecting
+	// skips
+	skipTo := initialWrites + 10
+	writeToFile(skipTo, expected[initialWrites:])
+
+	select {
+	case <-fullResultCh:
+	case <-time.After(10 * time.Duration(testutil.TestMultiplier()) * streamBatchWindow):
+		t.Fatalf("did not receive data: got %q", string(received))
+	}
+}
```

Comment:
These tests are pretty great. 

i would also add tests that verifies that are there no file descriptor leaks when the client closes the connection, or when the context is cancelled. also something we could do as part of end to end testing rather than here.

Reply:
Hmm I am not sure how you would do that in a unit test?

Reply:
using os.getPID to get the test processes's PID and verifying via runtime.exec of lsof -p <pid> that there are no open file handles to any alloc subdirectories will work though it will look ugly. We would have to make the test be linux only. Its much cleaner to do it in an integration test and examine it from the outside though. 

Thinking through this again, what I really want to see was if there's a way to assert that there aren't open sockets on the client in a weird state when streaming logs and there are errors or other conditions that cause that method to return. for example if someone accidentally deletes that defer conn.close() line can we catch that?